### PR TITLE
Add support for stats on timestamp type in Delta Lake

### LINKED ISSUE #21878: Delta Lake connector doesn't write min/max stats on `BOOLEAN`, `TIMESTAMP`, and `VARBINARY` types
```sql
CREATE TABLE test(a boolean, b TIMESTAMP(6), c varbinary, part VARCHAR) WITH (partitioned_by = ARRAY['part']);

INSERT INTO test VALUES 
(true, TIMESTAMP '2001-05-06 12:34:56.123456', x'12', 'a'), 
(false, TIMESTAMP '2001-05-06 12:34:56.567890', x'34', 'a');

SELECT *  FROM "test$partitions";
```

Related: https://github.com/trinodb/trino/pull/20993

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

Example of stats generated by Spark:
* transaction log has timestamp millis in stats, e.g. `2020-08-26T01:02:03.123`
* checkpoint has timestamp millis in stats, e.g. `2024-05-29T15:08:00.991Z`
* data file has timestamp micros, e.g. `2020-08-26T01:02:03.123987` with INT64

Both int96 and int64 are allowed in parquet files: https://github.com/delta-io/delta/blob/master/PROTOCOL.md#delta-data-type-to-parquet-type-mappings

Fixes #21878

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# Delta Lake
* TBD. ({issue}`21878`)
```

### CODE REVIEW

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeParquetStatisticsUtils.java [hunk 1]
```diff
@@ -137,7 +142,9 @@ public static Object jsonValueToTrinoValue(Type type, @Nullable Object jsonValue
         if (type == DateType.DATE) {
             return LocalDate.parse((String) jsonValue).toEpochDay();
         }
-        if (type == TIMESTAMP_MILLIS) {
+        if (type instanceof TimestampType timestampType) {
+            int precision = timestampType.getPrecision();
+            checkArgument(precision == 3 || precision == 6, "Unsupported precision: %s", precision);
             return Instant.parse((String) jsonValue).toEpochMilli() * MICROSECONDS_PER_MILLISECOND;
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeParquetStatisticsUtils.java:148]: `Instant.parse((String) jsonValue).toEpochMilli()` seems wrong for precision=6

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeParquetStatisticsUtils.java [hunk 2]
```diff
@@ -344,8 +356,8 @@ private static Optional<Object> getMin(Type type, Statistics<?> statistics)
             return Optional.of(new String(((BinaryStatistics) statistics).genericGetMin().getBytes(), UTF_8));
         }
 
-        if (type.equals(BOOLEAN)) {
-            // Boolean columns do not collect min/max stats
+        if (type.equals(BOOLEAN) || type.equals(VARBINARY)) {
+            // Boolean and varbinary columns do not collect min/max stats
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeParquetStatisticsUtils.java:360]: Let's extract varbinary changes into own commit, so they don't go as timestamp-related changes.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeParquetStatisticsUtils.java (hunk 1 above)
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeParquetStatisticsUtils.java:151]: Spark always writes timestamp stats with millisecond precision. Left a code comment and asking in https://delta-users.slack.com/archives/CJ70UCSHM/p1716970335747839

Comment (findepi) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeParquetStatisticsUtils.java:151]: for stats millis precision is fine (provided that we never take them as actual min/max)

however, the code here is truncating to millis, which is not correct for max value. it should round up

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TransactionLogParser.java [hunk 3]
```diff
@@ -119,7 +119,7 @@ private TransactionLogParser() {}
             .withResolverStyle(ResolverStyle.STRICT);
     public static final DateTimeFormatter JSON_STATISTICS_TIMESTAMP_FORMATTER = new DateTimeFormatterBuilder()
             .parseCaseInsensitive()
-            .appendValue(YEAR, 4, 10, SignStyle.NORMAL)
+            .appendValue(YEAR, 4, 10, SignStyle.EXCEEDS_PAD)
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TransactionLogParser.java:122]: The context: https://github.com/trinodb/trino/issues/22184

Comment (findepi) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TransactionLogParser.java:122]: can this change be separate commit (along with some test if any applicable)?

**Comment by ebyhr** [comment]
@findepi Addresed comments.

**Review by findepi — APPROVED** [review]

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeParquetStatisticsUtils.java [hunk 4]
```diff
@@ -366,6 +404,29 @@ private static Optional<Object> getMax(Type type, Statistics<?> statistics)
             return Optional.of(date.format(ISO_LOCAL_DATE));
         }
 
+        if (type instanceof TimestampType) {
+            if (statistics instanceof LongStatistics longStatistics) {
+                long epochMicros = longStatistics.genericGetMax();
+                long epochSeconds = floorDiv(epochMicros, MICROSECONDS_PER_SECOND);
+                int nanoAdjustment = floorMod(epochMicros, MICROSECONDS_PER_SECOND) * NANOSECONDS_PER_MICROSECOND;
+                Instant ts = Instant.ofEpochSecond(epochSeconds, nanoAdjustment);
+                Instant truncatedToMillis = ts.truncatedTo(MILLIS);
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeParquetStatisticsUtils.java]: Are you suggesting a conversion from `TimestampType.TIMESTAMP_MILLIS` to `ChronoUnit.MILLIS`. Can you share the code snippet?

[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeParquetStatisticsUtils.java]: Let me add a code comment instead. I didn't come up with a simple code for such a conversion.
